### PR TITLE
perf(mhc): tune SM120 medium prefill projection

### PR DESCRIPTION
**Pull request by voipmonitor**
## Result

Status: **implemented and qualified** for the hidden-size-4096 multi-head
connection (mHC) TF32 projection on SM120.

Projection batches from 2,304 through 3,583 rows use an M64/N24/K64 geometry
with eight K splits. Three pipeline stages cover 2,304 through 3,071 rows and
two stages cover 3,072 through 3,583 rows. The wide-CTA chunk geometry starts
at 3,584 rows for hidden size 4096.

## Technical reason

The scalar M16/N8/K256 geometry exposes too few cooperative thread arrays for
medium prefill batches. The qualified geometry increases SM occupancy while
letting one thread block own all 24 mix columns, which avoids loading the same
input tile once per eight-column output tile.

## Operation contract and compatibility

- The policy applies only to hidden size 4096 on the TF32 prefill projection.
- Hidden size 7168 retains the generic 4,096-row chunk threshold.
- Decode dispatch, scratch ownership, output layout, numeric formats, and
  public operation signatures are unchanged.
- Environment overrides remain available for every threshold and geometry
  field. Resolved geometry and thresholds are part of the compile key.
- The benchmark resolves geometry through the same policy function used by the
  production launcher.

## Correctness validation

Source base: B12X `master` at
`fc1d4b68f7a5b0cfdb88bf06abccd869f5c589d5`.

Hardware: physical GPU 0,
`GPU-d8438b2d-f000-a617-5dcc-0197ce0365a3`, NVIDIA RTX PRO 6000 Blackwell
Workstation Edition, PCIe Gen5 x16.

```bash
CUDA_VISIBLE_DEVICES=0 CUTE_DSL_ARCH=sm_120a \
  /opt/venv/bin/python -B -m pytest -q \
  tests/gemm/test_launch_custom_ops.py \
  -k "mhc_prefill_tf32_projection_geometry_boundaries or mhc_prefill_tf32_optimized_geometry_matches_reference_under_graph_replay or mhc_launch_ops_have_fake_dispatch"
```

Result: `5 passed, 10 deselected`.

The CUDA graph cases cover 2,304, 3,072, and 3,583 rows, mutate the input,
poison the reduction workspace, replay the captured operation, and compare
with `torch.nn.functional.linear`.

## Performance qualification

At 3,072 rows, five balanced process-level medians decreased from 261.07 µs on
`master` to 198.27 µs on the qualified implementation. The fixed-work latency
reduction is 24.05%; the equivalent projection-throughput increase is 31.67%.

The exact command, revisions, worktrees, GPU mode, CUTLASS/PTXAS map, fixed
cache manifests, cold and warm replay samples, correctness tolerances, and
ratio direction are recorded in
`docs/evidence/mhc_hidden4096_medium_prefill_sm120.md`.

This result qualifies the projection operation. It does not infer
end-to-end serving throughput from one kernel family.

## AI assistance disclosure

AI assistance was used to implement the policy and tests, perform correctness
and performance qualification, and prepare this pull-request description.

## Summary

- Tunes SM120 TF32 mHC prefill projection for hidden size 4096.
- Uses M64/N24/K64 geometry with eight K splits for 2,304–3,583 rows.
- Uses three pipeline stages for 2,304–3,071 rows and two stages for 3,072–3,583 rows.
- Keeps wide-CTA geometry for 3,584+ rows.
- Centralizes geometry selection and includes configurable geometry values in compile-cache keys.
- Keeps hidden size 7168 and decode dispatch unchanged.
- Updates the benchmark to use production geometry selection and adds raw timing output.
- Adds boundary and CUDA graph replay tests.

## Performance

At 3,072 rows, median latency decreases from 261.07 µs to 198.27 µs, a 24.05% reduction. Projection throughput increases by 31.67%.

## Validation

Five focused tests pass, including CUDA graph replay cases for 2,304, 3,072, and 3,583 rows.

### CODE REVIEW

**Comment by voipmonitor** [comment]
@coderabbitai review

**Comment by voipmonitor** [comment]
@coderabbitai review

File: docs/evidence/mhc_hidden4096_medium_prefill_sm120.md [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+| Revision | Geometry | Median samples, microseconds | Median |
+| --- | --- | --- | ---: |
+| Comparison | M16/N8/K256, one K split | 261.25, 263.09, 261.25, 260.02, 259.71 | 261.25 |
+| Qualified | M64/N24/K64, eight K splits, two stages | 198.27, 197.97, 196.74, 198.27, 198.27 | 198.27 |
+
+The fixed-work latency reduction is 24.11%. The equivalent throughput ratio is
+`261.25 / 198.27 = 1.317648`, or 31.76% higher projection throughput. The
+qualified output reported maximum projection error 0.0625 and projection RMSE
+0.000887; the comparison reported 0.0625 and 0.000902.
```
[the remark this answers is not in the document]
Comment (voipmonitor) [reply docs/evidence/mhc_hidden4096_medium_prefill_sm120.md]: Addressed in 168ca6bf and e5509cc9. The evidence now records both worktrees, physical GPU mode before and after collection, the CUTLASS DSL/PTXAS map, source-specific cache-manifest hashes, zero-warmup and preconditioned samples, 30 raw replay samples per arm, balanced arm order, correctness state, and the comparison/qualified ratio direction.

**Comment by voipmonitor** [comment]
@coderabbitai review

**Comment by voipmonitor** [comment]
@coderabbitai full review

**Comment by lukealonso** [comment]
Thanks @voipmonitor — this exposed a real policy gap: norm.mhc only exposed a fixed native config while the TF32 projection geometry lived behind import-time environment variables, so the profile generator could not discover this regime.

I have incorporated the M64/N24/K64 split-8 geometry as a typed, plan-owned tf32_tma backend and added a correctness-gated production CUDA-graph race across hidden sizes 4096/7168 and capacities through 8192. On pulsar’s RTX PRO 6000 Max-Q GPUs, the tuner independently selected the proposed M64, two-stage, split-8 plan at M=3072 (180.160 us, tied with M128), and selected the existing M192/M128 regimes at 3584/8192. At M=2304 it found M64 with two stages marginally faster than three stages (124.928 vs 125.632 us), which is exactly why this belongs in per-device profiles.

Closing this PR in favor of the profile-backed implementation prepared as commit 35065066. Thank you for finding the geometry and providing the benchmark evidence.
